File: benchmarks/serverSimulation/erlang/response.erl
-module(response).
-mode(compile).

% -compile(export_all).
-export([main/1]).

build(1) -> {nil, nil};
build(D) -> {build(D-1), build(D-1)}.

count({nil, nil}) -> 1;
count({L, R}) -> 1 + count(L) + count(R).

reaper(0) ->
  ok;
reaper(NClients) ->
  receive done -> reaper(NClients-1) end.

final(0) ->
  main ! done;
final(NRequests) ->
  receive {done, Id, Count, Start} -> ok end,
  Diff = erlang:monotonic_time() - Start,
  MS = erlang:convert_time_unit(Diff, native, micro_seconds),
  io:format("(~p, ~p, ~p)~n", [Id,Count, MS]),
  final(NRequests-1).

server() ->
  receive {Id, Depth, Final, Start} -> ok end,
  Final ! {done, Id, count(build(Depth)), Start},
  server().

client(Server, Id, NRequests, Final) ->
  [Server ! {Id, Depth, Final, erlang:monotonic_time()}
   || I <- lists:seq(0, NRequests-1), Depth <- [2*I + 6]].

main(Args) ->
  register(main, self()),
  [NServers, NClients, NRequests] =
    lists:map(fun erlang:list_to_integer/1, Args),
  Servers = list_to_tuple(
              [spawn(fun server/0) || _ <- lists:seq(1, NServers)]
             ),
  [spawn(fun() ->client(
                   element((I rem NServers)+1, Servers),
                   I,
                   NRequests,
                   spawn(fun() -> final(NRequests) end))
         end
        )
   || I <- lists:seq(0, NClients-1)],
  reaper(NClients),
  ok.

File: benchmarks/serverSimulation/akka/response.scala
import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Props
import akka.util.Timeout
import akka.pattern.ask
import com.typesafe.config.Config
import com.typesafe.config.ConfigFactory
import com.typesafe.config.ConfigValueFactory
import com.typesafe.config.impl.ConfigLong
import scala.language.postfixOps
import scala.concurrent.duration._

final class Tree(left:Tree, right:Tree) {
  def count:Int = if (left eq null) 1 else 1 + left.count + right.count
}

object Tree {
  def apply(depth:Int):Tree = {
    if (depth > 1) new Tree(Tree(depth-1), Tree(depth-1))
    else new Tree(null, null)
  }
}

class Reaper(var countdown: Int) extends Actor {
  def receive: Actor.Receive = {
    case "done" => {
      countdown -= 1
      if (countdown == 0) {
        System.exit(0)
      }
    }
  }
}

case class Request(id:Int, depth:Int, next:ActorRef, start:Long)
case class Response(id:Int, count:Int, start:Long)

class Final(var n_requests:Int) extends Actor {
  def receive: Actor.Receive = {
    case Response(id, count, start) => {
      val end = System.nanoTime()
      println(s"($id, $count, ${(end-start)/1000})")
      n_requests -= 1
      if (n_requests == 0) {
        context.actorSelection("/user/reaper") ! "done"
      }
    }
  }
}

class Server() extends Actor {
  def receive: Actor.Receive = {
    case Request(id, depth, next, start) =>
      next ! Response(id, Tree(depth).count, start)
  }
}

class Client(id:Int, server:ActorRef, n_requests:Int, next:ActorRef)
    extends Actor {

  override
  def preStart = {
    for (i <- 0 until n_requests) {
      val depth = i*2 + 6
      val start = System.nanoTime()
      server ! Request(id, depth, next, start)
    }
  }

  def receive: Actor.Receive = {
    case _ => println("sth")
  }
}

object Main {
  // val config =
  //   ConfigFactory.parseString(
  //       """
  //       akka{actor{default-dispatcher{fork-join-executor{parallelism-max = 1}}}}
  //       """
  //   )
  //   .withFallback(ConfigFactory.defaultReference(this.getClass.getClassLoader))

  // val system = ActorSystem("system",config)

  val system = ActorSystem("system")

  def main(args: Array[String]) {
    if (args.size < 3) {
      println("program <#server> <#client> <#requests>")
      System.exit(-1)
    }
    val n_servers = args(0).toInt
    val n_clients = args(1).toInt
    val n_requests = args(2).toInt

    val servers = (
        for (i <- 0 until n_servers)
          yield system.actorOf(Props(classOf[Server]), "server_" + i)
        )

    system.actorOf(Props(classOf[Reaper], n_clients), "reaper")

    for (i <- 0 until n_clients) {
      val next = system.actorOf(Props(classOf[Final], n_requests), "final_" + i)
      system.actorOf(Props(classOf[Client],
            i, servers(i%n_servers), n_requests, next), "client_" + i)
    }
  }
}
